File: data-structures/queue/typescript/queue.ts
export class Node<T> {
	next: Node<T> | null = null;
	constructor(public value: T) {}
}

export class Queue<T> {
	rear: Node<T> | null = null;
	front: Node<T> | null = null;
	length: number = 0;

	enqueue(value: T) {
		const newNode = new Node(value);

		if (this.length === 0) {
			this.rear = newNode;
			this.front = newNode;
		} else {
			this.rear!.next = newNode;
			this.rear = newNode;
		}

		this.length++;
	}

	dequeue(): T | null {
		if (this.length === 0) {
			return null;
		}

		const removedNode = this.front;
		this.front = this.front!.next;

		removedNode!.next = null;

		this.length--;
		return removedNode!.value;
	}

	size(): number {
		return this.length;
	}

	peek(): T | null {
		return this.front ? this.front.value : null;
	}

	printQueue(): T[] {
		const queueList = [];
		let currentNode = this.front;

		while (currentNode) {
			queueList.push(currentNode.value);
			currentNode = currentNode.next;
		}

		return queueList;
	}
}

File: data-structures/queue/python/queue.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class Queue:
    def __init__(self):
        self.rear = None
        self.front = None
        self.length = 0

    def enqueue(self, value):
        new_node = Node(value)

        if self.length == 0:
            self.rear = new_node
            self.front = new_node
        else:
            self.rear.next = new_node
            self.rear = new_node

        self.length += 1

    def dequeue(self):
        if self.length == 0:
            return None

        removed_node = self.front
        self.front = self.front.next

        removed_node.next = None

        self.length -= 1
        return removed_node.value

    def size(self):
        return self.length

    def peek(self):
        return self.front.value if self.front else None

    def print_queue(self):
        queue_list = []
        current_node = self.front

        while current_node:
            queue_list.append(current_node.value)
            current_node = current_node.next

        return queue_list
